compiler/rustc_arena/src/lib.rs

```diff
@@ -114,6 +117,102 @@ impl<T> Default for TypedArena<T> {
     }
 }
 
+#[inline]
+unsafe fn write_from_iter<'a, T, I: Iterator<Item = T>>(
+    mut iter: I,
+    len: usize,
+    mem: *mut T,
+) -> &'a mut [T] {
+    let mut i = 0;
+    // Use a manual loop since LLVM manages to optimize it better for
+    // slice iterators
+    loop {
+        let value = iter.next();
+        if i >= len || value.is_none() {
+            // We only return as many items as the iterator gave us, even
+            // though it was supposed to give us `len`
+            return slice::from_raw_parts_mut(mem, i);
+        }
+        ptr::write(mem.add(i), value.unwrap());
+        i += 1;
+    }
+}
+
+trait IterExt<T> {
+    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T];
+}
+
+impl<I, T> IterExt<T> for I
+where
+    I: IntoIterator<Item = T>,
+{
+    #[inline]
+    default fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
+        let mut vec: SmallVec<[_; 8]> = self.into_iter().collect();
+        if vec.is_empty() {
+            return &mut [];
+        }
+        // Move the content to the arena by copying it and then forgetting
+        // the content of the SmallVec
+        unsafe {
+            let len = vec.len();
+            let start_ptr = arena.alloc_raw_slice(len);
+            vec.as_ptr().copy_to_nonoverlapping(start_ptr, len);
+            vec.set_len(0);
+            slice::from_raw_parts_mut(start_ptr, len)
+        }
```

Comment:
Is there a reason this is duplicating the specialization for SmallVec? I would expect us to just forward to that here.